In [1]:
from fourier import fourier2d, ifourier2d
from lens import defines
import numpy as np

# A rough approximation of the lens

Rather then modeling the actual ASML lenses we simplify:
- The image plane is homogeneous
- A single Pupil plane exists
- No magnification
- Homogeneous incoming light / electric field
- Several other properties, no doubt... 

In [2]:
wavelength = 193e-9
NA = 1.35

$$
P = \left\{ (f_x, f_y) : \sqrt{f_x^2 + f_y^2}< \frac{NA}{\lambda} \right\}
$$

$$
f_x = \frac{n x'}{z \lambda}
$$

$$
f_x = \frac{n y'}{z \lambda}
$$

So $x'^2 + y'^2 < \frac{NA \cdot z}{n}$


For $n$ it is not entirely clear whether it is the refractive index of everything between object and image, or whether the immersion fluid under the lens already justifies the usage of $n_{water}$. 

In [3]:
z = .5
n = defines.N_WATER_AT_193_NM

In [4]:
np.round((NA / wavelength) / (n / (z*wavelength)), 3)

0.47

So for our dimensions $x'^2 +y'^2 < .47 m$, let's continue with this simplification for the time being.

# The pupil plane

Let

$
\left( \begin{array}{ll}
f_x  \\
f_y
\end{array} \right) = \frac{NA \cdot R}{ \lambda}
\left( \begin{array}{ll}
\cos(\phi) \\
\sin(\phi)
\end{array} \right)$

Let the object be modeled as an identity function: $\mathbb{1}_{A}$,

$$
\int_{-\infty}^{\infty} \int_{-\infty}^{\infty} \mathbb{1}_A(x, y) e^{-2 \pi i (f_x x+f_y y)} dx dy
$$

The pupil plane has a cut of frequency as we saw. This is applied via multiplication with the pupil function:
$$
P(f_x, f_y) = \left\{ \begin{array}{ll} 
1 & : \sqrt{f_x^2 + f_y^2} < \frac{NA}{\lambda} \\
0 & : otherwise
\end{array}\right.
$$

# Including aberrations

We also express aberrations in the pupil plane. For this we add the following additional factor:
$$
P(f_x, f_y) e^{i 2 \pi W(R, \phi)},
$$

where $W$ corresponds to the wavefront aberrations.